src/ImageSharp.Drawing/Processors/FillShapeProcessor.cs

```diff
@@ -61,77 +62,398 @@ protected override void OnApply(ImageBase<TColor> source, Rectangle sourceRectan
             minY = Math.Max(0, minY);
             maxY = Math.Min(source.Height, maxY);
 
-            // Reset offset if necessary.
-            if (minX > 0)
-            {
-                startX = 0;
-            }
+            ArrayPool<Vector2> arrayPool = ArrayPool<Vector2>.Shared;
 
-            if (minY > 0)
-            {
-                polyStartY = 0;
-            }
+            int maxIntersections = this.poly.MaxIntersections;
 
             using (PixelAccessor<TColor> sourcePixels = source.Lock())
-            using (IBrushApplicator<TColor> applicator = this.fillColor.CreateApplicator(sourcePixels, rect))
+            using (BrushApplicator<TColor> applicator = this.fillColor.CreateApplicator(sourcePixels, rect))
             {
                 Parallel.For(
                 minY,
                 maxY,
                 this.ParallelOptions,
                 y =>
                 {
-                    int offsetY = y - polyStartY;
+                    var buffer = arrayPool.Rent(maxIntersections);
```

Comment:
the buffer is used in parallel... it needs to be rented inside the parallel loop otherwise multiple threads will be competing over the same buffer.